### ISSUE TITLE
DenyHosts with Exim4 - No mail is sent

### ISSUE THREAD

**Issue by ZerooCool** [denyhosts/denyhosts#168]
DenyHosts with Exim4 - No mail is sent

I do not know if this problem is related to the two issues here, but, still no mails, locally, for root, nor, remotely. #89 #93

Here I installed Exim4.
You can consult the stream to see the problem and observe the method.

_Available 14 days_ : https://www.twitch.tv/videos/629690890

**Comment by ZerooCool** [comment]
A single email is sent to me here, possibly following the restart of the Exim4 service last night.
I had the same thing recently, on my other production server, a single mail sent because, never arrived correctly, and, is redistributed by Mailer-Daemon.

This is all the more surprising since in my Exim configuration, all system emails should be sent to root, and root should send to a valid email address.

It works very well with Fail2ban.
It does not work with DenyHosts.

```
From : Mail Delivery System <Mailer-Daemon@My-VPS>
Sujet : Mail delivery failed: returning message to sender
For : nobody@localhost

------------------------------------------------

This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a permanent error. The following address(es) failed:

  my-real-adress@my-domain.ext
    host mx1.ovh.net [188.165.47.122]
    SMTP error from remote mail server after RCPT TO:<my-real-adress@my-domain.ext>:
    504 5.5.2 <nobody@localhost>: Sender address rejected:
    need fully-qualified address

Reporting-MTA: dns; My-VPS

Action: failed
Final-Recipient: rfc822;my-real-adress@my-domain.ext
Status: 5.0.0
Remote-MTA: dns; mx1.ovh.net
Diagnostic-Code: smtp; 504 5.5.2 <nobody@localhost>: Sender address rejected: need fully-qualified address


ForwardedMessage.eml

Sujet : DenyHosts Report
--
De : DenyHosts <nobody@localhost>
Date : 24/05/2020 à 00:00
```

**Comment by ZerooCool** [comment]
Note, this mails was send with denyhosts 2.10-2
(Bounce from the server)

```
ADMIN_EMAIL = root@localhost
#SMTP_HOST = localhost
#SMTP_PORT = 25
#SMTP_FROM = DenyHosts <nobody @ localhost>
```

By default, the FROM is not specified since commented.
It is certainly nobody@localhost which is used.
The following error is displayed:
```
SMTP error from remote mail server after RCPT TO: <real@email.ext>:
504 5.5.2 <nobody@localhost>: Sender address rejected:
need fully-qualified address
```

I try to uncomment and I use my administrator address.
The recipient is root@localhost, so the message will be sent to real@address.ext.

```
# For
ADMIN_EMAIL = root@localhost
# From
SMTP_FROM = DenyHosts Email Sender <real@address.ext>
SMTP_SUBJECT = DenyHosts Report VPS
```

**My assumption is that using the nobody@localhost address by default is the problem.**

In addition, I noticed in the bounce, a reference to SPF, like what nobody@localhost is not present in the SPF.
This is a double problem!
1- I did not correctly finalize the EXIM4 configuration step, with regard to:
- SPF
- DKIM
- DMARC

2- But, nobody@localhost will not be added to the SPF anyway, it is therefore imperative to configure a valid sender.

It is therefore necessary to inform:
- The target who will receive the mail: root@localhost
- The sender of the mail, who will not be nobody@localhost

Now, I ask myself the question, if Fail2ban for example, can send an email, without specifying the sender (I think) and only specify the administrator (The recipient is root@localhost) then DenyHosts should be able to do it too.

**My question would be: Is it possible to use something other than nobody @ localhost by default, to modify the initial configuration ?**

I look the header From Fail2ban : 

Subject: [Fail2Ban] sshd: banned 222.186.42.137 from SERVEUR
Date: Sun, 14 Jun 2020 03:04:34 +0200
From: Fail2Ban <real@adresse-number-1.ext>
To: real@adresse-number-1.ext

Use 2x the same address, for me, it's good. Admin from serveur, send the mail to the admin.


If I look at my notes, I confirm that for Fail2ban I only have to fill in the recipient's email!
On the other hand, Fail2ban asks to inform the MTA : 
https://wiki.visionduweb.fr/index.php?title=Installer_et_utiliser_Fail2ban#Recevoir_un_mail_en_cas_de_ban

It is explained that to send mails with Fail2ban, it is necessary to replace the MTA mail by sendmail
```
# Replace mta = mail with:
mta = sendmail
```
So I'm not sure who sends my emails from Fail2ban? Sendmail? Exim?

It seems to be sendmail, but, I don't think I have configured it on sendmail, so the installation of Exim takes over.

The only configuration I did for sendmail, is the PHP configuration, for which I ask the sendmail_path to use Exim4.
`sendmail_path = /usr/sbin/ exim4 -t -i -F real@adresse-number-1.ext -f real@adresse-number-1.ext`

If I send a test email, on the command line, with sendmail :
`(echo "Subject: Content of mail2"; echo "Subject of mail") | sudo sendmail my@email.ext`

```
Received: from root by VPS with local (Exim 4.93)
(envelope-from <my@email.ext>)
id 1jkIC5-002MTP-Jf
for my@email.ext; Sun, 14 Jun 2020 04:16:21 +0200
Subject: Content of mail2
Message-Id: <E1jkIC5-002MTP-Jf@VPS>
From: root <my@email.ext>
```

So sendmail uses Exim well and Exim sends email well.

In conclusion :
By default, Denyhosts must use sendmail ?
Shouldn't the default value of SMTP_FROM be used, with a value of nobody@localhost ? (especially if SMTP_FROM is commented)

**Comment by ZerooCool** [comment]
As I wrote, I tried to uncomment the FROM and put a valid email address, but, it still doesn't work.
`SMTP_FROM = DenyHosts <real@address.ext>`

Now, i have use a new syntaxe, normally, it’s even worse to do so, so, really, it won’t be anything better :
`SMTP_FROM = DenyHosts real@address.ext`

I receive, at least for one of my servers, the bounce mails, that is to say, the message has not been delivered, and, it is then the administrator who receives the bounces.

Now, as I said, Exim is not perfectly configured for me, and, the SPF should authorize the email used to send messages, except, I may not have done it correctly. (I'm almost sure I didn't do it.)

Despite everything, fail2ban, it manages to send messages, and it is based on Exim, therefore, the part which is already configured in a rudimentary and automatic way, and, it works.

I continue slowly, to investigate, to see if I manage to obtain a functional behavior, in the days to come surely, by then configuring the SPF correctly.

**Comment by josev814** [comment]
I recently rewrote the e-mail functionality into it's own file/class.  One thing that I've noticed, is that by default the denyhosts.conf file sets the SMTP_SSL setting to yes, but it's using port 25 for localhost.  Could you try adjusting the setting SMTP_SSL setting to no, and see if that resolved your issue?  I'm still testing out the email class, but from what I'm seeing that should resolve your issue, unless you are wanting to use SSL.  Then change the port from 25 to your SSL/TLS port that you're wanting to use.

**Comment by ZerooCool** [comment]
I test that :
`SMTP_SSL = no`

I'm waiting for the next email in my box.

On the other hand, if I understand correctly, the sending of the message will not be protected?
Not very sure of what I do by passing SMTP_SSL to no.
If it is EXIM which sends the messages, EXIM him, uses a certificate which should be used to encrypt the exchanges, but, I suppose that there could remain a part of the communication which will not be encrypted, during the exchange between Denyhosts and Exim?
I have absolutely no idea, and, I would point out that in any case, I am speculating, because here, there is always a question of the obsolete version 2.10, whereas I had better just advance my feedback on version 3 directly, something that I will do it soon (switch all my versions to version 3.)

**Comment by josev814** [comment]
The SMTP_SSL version is pertaining to version 3.1.x.  Version 2.10 doesn't
have the SMTP SSL settings.
